gems/smithy-client/spec/smithy-client/plugins/resolve_auth_spec.rb

```diff
@@ -0,0 +1,162 @@
+# frozen_string_literal: true
+
+require_relative '../../spec_helper'
+
+require 'smithy-client/plugins/resolve_auth'
+
+module Smithy
+  module Client
+    module Plugins
+      describe ResolveAuth do
+        let(:shapes) { ClientHelper.sample_shapes }
+        let(:sample_client) { ClientHelper.sample_client(shapes: shapes) }
+
+        let(:client_class) do
+          client_class = sample_client.const_get(:Client)
+          client_class.clear_plugins
+          client_class.add_plugin(sample_client::Plugins::Endpoint)
+          client_class.add_plugin(AnonymousAuth)
+          client_class.add_plugin(Protocol)
+          client_class.add_plugin(ResolveAuth)
+          client_class.add_plugin(StubResponses)
+          client_class
+        end
+
+        let(:client) { client_class.new(stub_responses: true) }
+
+        it 'adds an :auth_resolver option to config' do
+          expect(client.config).to respond_to(:auth_resolver)
+        end
+
+        it 'adds scheme ids to auth scheme hash' do
+          client_class.add_plugin(HttpApiKeyAuth)
+          client_class.add_plugin(HttpBasicAuth)
+          client_class.add_plugin(HttpBearerAuth)
+          client_class.add_plugin(HttpDigestAuth)
+          client
+          expect(client.config.auth_schemes['smithy.api#noAuth']).to equal(:anonymous_provider)
+          expect(client.config.auth_schemes['smithy.api#httpApiKeyAuth']).to equal(:http_api_key_provider)
+          expect(client.config.auth_schemes['smithy.api#httpBasicAuth']).to equal(:http_login_provider)
+          expect(client.config.auth_schemes['smithy.api#httpBearerAuth']).to equal(:http_bearer_provider)
+          expect(client.config.auth_schemes['smithy.api#httpDigestAuth']).to equal(:http_login_provider)
+        end
+
+        context 'resolving auth' do
+          it 'resolves auth for anonymous auth' do
+            handler = ResolveAuth::Handler
+            expect_any_instance_of(handler).to receive(:try_load_auth_scheme).and_wrap_original do |method, *args|
+              resolved_auth = method.call(*args)
+              expect(resolved_auth[:scheme_id]).to equal('smithy.api#noAuth')
+              expect(resolved_auth[:identity]).to be_a(Identities::Anonymous)
+              resolved_auth
+            end
+            client.operation
+          end
+
+          it 'resolves auth for http api key auth' do
+            shapes['smithy.ruby.tests#SampleClient']['traits']['smithy.api#httpApiKeyAuth'] = {}
+            client_class.add_plugin(HttpApiKeyAuth)
+            handler = ResolveAuth::Handler
+            expect_any_instance_of(handler).to receive(:try_load_auth_scheme).and_wrap_original do |method, *args|
```

Comment:
Rather than do this - I think you just want to do something like
```
resp = client.operation
expect(resp.context.auth[:scheme_id]) ...
etc
```
This is more behavior driven and not implementation (handler wrapping)